### DISCUSSION TITLE
Flycast core crash after retro_init() in my libretro implementation

### DISCUSSION THREAD

**Discussion by deusX1** [flyinghead/flycast#1370] [Q&A] [answered]
Hello everyone,
I'm currently working on a java libretro implementation made just for fun that later I will add on github, when it's ready. 
The most important parts of the api have been implemented, so it can actually init and run almost every cores without issues.

However, I'm having an issue during the initialization of the Flycast core: just a few secs after returning from retro_init() it crashes with error. 

I have to add that not all the environments commands are implemented, and also I'm currently trying to figure out how to correctly print messages from the log callback in java, so I cannot see what type of error is at this time. 

This is a log of what's happening:

-> loadCore()
retro_api_version= 1
retro_sys_info[lib_name]= Flycast
retro_sys_info[lib_version]= 195f4010
retro_sys_info[valid_extensions]= chd|cdi|elf|cue|gdi|lst|bin|dat|zip|7z|m3u
retro_sys_info[need_fullpath]= true
retro_sys_info[block_extract]= false

[CMD 52] RETRO_ENVIRONMENT_GET_CORE_OPTIONS_VERSION
- ignored

[CMD 16] RETRO_ENVIRONMENT_SET_VARIABLES
- set_frontend | variable -> (reicast_region, Region; USA|Japan|Europe|Default)
....
....
- set_frontend | variable -> (reicast_vmu4_screen_opacity, VMU Screen 4 Opacity; 100%|10%|20%|30%|40%|50%|60%|70%|80%|90%)

[CMD 69] RETRO_ENVIRONMENT_SET_CORE_OPTIONS_UPDATE_DISPLAY_CALLBACK
- ignored

[CMD 35] RETRO_ENVIRONMENT_SET_CONTROLLER_INFO
- set_frontend | new controller-info -> subclasses = 13
- set_frontend | controller-description -> desc = Controller id = 1
- set_frontend | controller-description -> desc = Arcade Stick id = 1025
- set_frontend | controller-description -> desc = Keyboard id = 3
- set_frontend | controller-description -> desc = Mouse id = 2
- set_frontend | controller-description -> desc = Light Gun id = 4
- set_frontend | controller-description -> desc = Twin Stick id = 513
- set_frontend | controller-description -> desc = Saturn Twin-Stick id = 769
- set_frontend | controller-description -> desc = Pointer id = 6
- set_frontend | controller-description -> desc = Maracas id = 1281
- set_frontend | controller-description -> desc = Fishing Controller id = 1537
- set_frontend | controller-description -> desc = Pop'n Music id = 1793
- set_frontend | controller-description -> desc = Race Controller id = 2049
- set_frontend | controller-description -> desc = Densha de Go! id = 2305
....
....

[CMD 27] RETRO_ENVIRONMENT_GET_LOG_INTERFACE
- set_core | log-interface callback
::[INFO] - %s

[CMD 28] RETRO_ENVIRONMENT_GET_PERF_INTERFACE
- ignored

[CMD 10] RETRO_ENVIRONMENT_SET_PIXEL_FORMAT
- set-frontend | pixel_format = XRGB8888

[CMD 12] RETRO_ENVIRONMENT_SET_KEYBOARD_CALLBACK
- ignored

[CMD 57] RETRO_ENVIRONMENT_GET_DISK_CONTROL_INTERFACE_VERSION
- ignored

[CMD 13] RETRO_ENVIRONMENT_SET_DISK_CONTROL_INTERFACE
- ignored

[CMD 31] RETRO_ENVIRONMENT_GET_SAVE_DIRECTORY
- set_core | save-dir = /Users/deus/Development/Projects/VisualStudio CODE/Libretro/data/save

::[WARN] - %s
::[WARN] - %s
::[INFO] - %s

-> loadCore() done!

::[ERROR] - %s

-> A fatal error has been detected by the Java Runtime Environment:
-> SIGSEGV (0xb) at pc=0x0000000118651268, pid=8892, tid=259
---------------------------------------------------------------------

So, what I am asking, are the RETRO_ENVIRONMENT* commands not yet implemented (ignored) mandatory for Flycast to work correctly or there is something else I'm not aware of?
If can help, I'm developing from an arm64 macos.

I also tried to study the flycast source, within my limits, but without success.

Any help will be very much appreciated, thank you in advance!

**Comment by flyinghead** [comment]
Flycast needs hardware rendering support (OpenGL, Vulkan or DirectX11) and it relies on RETRO_ENVIRONMENT_SET_HW_RENDER (CMD 14) to set the parameters for establishing the graphics API context.

**Reply by deusX1** [reply]
I already have an OpenGL context up and running (using a java opengl binding) currently used for other cores, but I have yet to implement CMD 14 for supporting hardware rendering cores. However CMD 14 is not sent in retro_init() from what I see, so I suppose my problem should be somewhere else? In the meanwhile I'll try to implement some missing environments commands and will fix the log printing to see exactly what is happening, then I will post updates here. Thank you.

**Reply by flyinghead** [reply]
> just a few secs after returning from retro_init() it crashes with error.

Flycast doesn't start any thread in retro_init() so it would mean the frontend itself is crashing.
In any case, debugging and getting a stacktrace when the crash happens would help.

**Reply by deusX1** [reply]
ok, error is:
::[ERROR] - core/linux/common.cpp:69 E[COMMON]: SIGSEGV @ 0x116e55948 invalid access to 0xc
I get this error only if I set the log callback. Probably something on the java/jna side is messing up things.. thank you anyway.

**Reply by flyinghead** [reply, marked as answer]
This is some (almost) null pointer exception.
Note that the flycast core installs a SIGSEGV fault handler so it will print out messages to the log when catching an unexpected one (as is the case here). However that doesn't mean the exception originates from flycast code.
